### PR TITLE
Add pipeline error for passing the same step object

### PR DESCRIPTION
**Pull request by kamalesh0406**
## Pre-requisites
Please ensure you have done the following:
- [x] I have read the **CONTRIBUTING.md** document.
- [x] If my change requires a change to the documentation, I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Describe changes
Briefly describe the changes you have introduced.

I have added a new check to the _verify_arguments in base_pipelines for similar objects being passed into different arguments as mentioned in #267. The only issue that is unresolved now is the test_base_pipeline.py, which is breaking because of this new error function.

### CODE REVIEW

**Review by htahir1 — CHANGES_REQUESTED** [review]
This is great! Again very well done. I only have one comment to change the error message!

**Comment by htahir1** [comment]
Ah a test now fails as well! I think we need to edit the test to accommodate for this PR

**Comment by kamalesh0406** [comment]
Yup the tests for pipelines all pass the same function to the arguments twice. So that is breaking with this new error.

**Comment by htahir1** [comment]
@kamalesh0406   Would you edit the tests or shall I do it? Btw I would change the error message to:

f"Same argument type (`{type(step)}`) used for arguments {key} and {prev_key}. Step objects should be unique for each argument." 

Because that would cover the class API as well as the functional API

**Comment by kamalesh0406** [comment]
I am trying to modify the error and I will update the error message as well.

**Comment by kamalesh0406** [comment]
I have added the tests and while doing so I caught a bug in the code I wrote previously. I found that the code only identified similar functions if they had the keyword argument evaluator= used in the function call and they didn't detect repeated arguments when the argument type was not specified. I have corrected this issue as well in the latest commit.

File: src/zenml/pipelines/base_pipeline.py [hunk 1]
```diff
@@ -181,8 +191,18 @@ def _verify_arguments(self, *steps: BaseStep, **kw_steps: BaseStep) -> None:
                     f"a pipeline."
                 )
 
+            step_ref_func = type(step)
+
+            if step_ref_func in step_refs:
+                step_refs[step_ref_func]
```
Comment (htahir1) [inline src/zenml/pipelines/base_pipeline.py:197]: In this case, wouldnt it make sense to have the error like you had before: `f"Same argument type ({type(step)}) used for arguments {key} and {prev_key}. Step objects should be unique for each argument."`

Comment (htahir1) [inline src/zenml/pipelines/base_pipeline.py:197]: This is a hanging reference

File: src/zenml/pipelines/base_pipeline.py [hunk 2]
```diff
@@ -158,9 +159,18 @@ def _verify_arguments(self, *steps: BaseStep, **kw_steps: BaseStep) -> None:
...
+            if step_ref_func in step_refs:
+                raise PipelineInterfaceError(
+                    f"Step object (`{type(step)}`) has been used twice. Step "
+                    f"objects should be unique for each argument."
+                )
+
             key = input_step_keys[i]
             step.pipeline_parameter_name = key
             combined_steps[key] = step
+            step_refs[step_ref_func] = str(i)
```
Comment (htahir1) [inline src/zenml/pipelines/base_pipeline.py:173]: Do we need to coerce the `i` to a string?

**Review by htahir1 — COMMENTED** [review]
Thanks for the latest changes. I had a few comments. @schustmi should definitely take a look at the changed tests, as he wrote them initially. 

Also we have one merge conflict that would be cool to resolve @kamalesh0406 ? Other than that looks good!

File: src/zenml/pipelines/base_pipeline.py [hunk 3]
```diff
@@ -147,6 +147,7 @@ def _verify_arguments(self, *steps: BaseStep, **kw_steps: BaseStep) -> None:
             )
 
         combined_steps = {}
+        step_refs: Dict[type, str] = {}
```
Comment (schustmi) [inline src/zenml/pipelines/base_pipeline.py:150]: I think a set is enough in this case as we don't actually use the dict values as far as I can see. I'd also try to be a little be more explicit with the actual type annotation of the set, in our case it can only contain `BaseStep` subclasses: 
```python
step_classes = Set[Type[BaseStep]] = set()
```

File: src/zenml/pipelines/base_pipeline.py [hunk 4]
```diff
@@ -158,9 +159,18 @@ def _verify_arguments(self, *steps: BaseStep, **kw_steps: BaseStep) -> None:
                     f"a pipeline."
                 )
 
+            step_ref_func = type(step)
```
Comment (schustmi) [inline src/zenml/pipelines/base_pipeline.py:162]: I think `step_class` is clearer and covers both the functional and class-based API. Maybe we can even define it right after the for-loop and use it inside the previous error message (which also calls `type(step)`?

File: tests/pipelines/test_base_pipeline.py [hunk 5]
```diff
@@ -44,46 +44,55 @@ def some_pipeline(step_):
 
 
 def test_initialize_pipeline_with_args(
-    unconnected_two_step_pipeline, empty_step
+    unconnected_two_step_pipeline, multiple_empty_steps
 ):
     """Test that a pipeline can be initialized with args."""
     with does_not_raise():
-        unconnected_two_step_pipeline(empty_step(), empty_step())
+        empty_step_1, empty_step_2, extra_steps = multiple_empty_steps()
```
Comment (schustmi) [inline tests/pipelines/test_base_pipeline.py:51]: Really like the idea with the multiple steps fixture, just one small suggestion: We could use an underscore if we don't need the extra steps, e.g. `empty_step_1, empty_step_2, _ = multiple_empty_steps()`. (Similar in the remaining tests)

**Review by schustmi — CHANGES_REQUESTED** [review]
Really like the implementation idea and tests, just added a few naming/type annotation suggestions

File: src/zenml/pipelines/base_pipeline.py (hunk 1 above)
Comment (kamalesh0406) [reply src/zenml/pipelines/base_pipeline.py:197]: Yup, but the issue is in the case when the user calls the pipeline function like this trainer(evaluator(), evaluator_2 = evaluator()) we won't be able to return the key.

File: src/zenml/pipelines/base_pipeline.py (hunk 2 above)
Comment (kamalesh0406) [reply src/zenml/pipelines/base_pipeline.py:173]: Not really I wanted to do it since it was a string as suggested below I will move it to a set rather than a Dict.

File: tests/pipelines/test_base_pipeline.py (hunk 5 above)
Comment (kamalesh0406) [reply tests/pipelines/test_base_pipeline.py:51]: okay, that is better, I will change them to _ .

File: src/zenml/pipelines/base_pipeline.py (hunk 3 above)
Comment (kamalesh0406) [reply src/zenml/pipelines/base_pipeline.py:150]: okay, even I had second thoughts about this, I will change it in the next commit.

File: src/zenml/pipelines/base_pipeline.py (hunk 4 above)
Comment (kamalesh0406) [reply src/zenml/pipelines/base_pipeline.py:162]: yup sorry I will change this as well. I couldn't think of a good variable name initially.

File: src/zenml/pipelines/base_pipeline.py (hunk 1 above)
Comment (htahir1) [reply src/zenml/pipelines/base_pipeline.py:197]: Yes for this case (args) we don't need the full message but for the other case (kwargs) we should have it  I think

Comment (kamalesh0406) [reply src/zenml/pipelines/base_pipeline.py:197]: Okay, I will try to see how to do it. It would mean that I have to initialize a new dict for the kwargs and check if the function is repeated there first and then check for the reference in the args.

**Comment by kamalesh0406** [comment]
I have resolved all the comments above in the previous commit and I have added two additional tests, one to test with repeated kwargs and the other to test with a mix of repeated args and kwargs so that we do not miss out on any edge cases.

**Review by htahir1 — COMMENTED** [review]
> I have resolved all the comments above in the latest commit and I have added two additional tests, one to test with repeated kwargs and the other to test with a mix of repeated args and kwargs so that we do not miss out on any edge cases.

@kamalesh0406 Awesome - It all seems in order now from my side. We just need to merge main and resolve the merge conflicts!

**Comment by kamalesh0406** [comment]
Okay, I cannot actually view the merge conflicts because github seems to be allowing it only for users with write access.

**Comment by htahir1** [comment]
@kamalesh0406 can you just try and merge main locally and see them

**Comment by kamalesh0406** [comment]
I thought the merge conflict was purely due to a change in the folder structure but it seems that some of the code needs to be changed as well. I am not able to recreate this on my main, can you look into the conflicts and merge the pull request? Thanks!

**Comment by htahir1** [comment]
@kamalesh0406 Alright i resolved them and pushed to main. Thank you so much for your contribution!
